pageserver/src/relish_storage/storage_sync.rs

```diff
@@ -0,0 +1,1417 @@
+//! NOTE: No backpressure or eviction is implemented for tasks that always fail, it will be improved later.
+//!
+//! Synchronization never removes any local from pageserver workdir or remote files from the remote storage: the files from previous
+//! uploads that are not mentioned in the new upload lists, are still considered as part of the corresponding image.
+//! When determining which files to upload/download, the local file paths (for remote files, that is the same as their download destination) is compared,
+//! and two files are considered "equal", if their paths match. Such files are uploaded/downloaded over, no real contents checks are done.
+//! NOTE: No real contents or checksum check happens right now and is a subject to improve later.
+//!
+//! After the whole is downloaded, [`crate::tenant_mgr::register_relish_download`] function is used to register the image in pageserver.
+//!
+//! When pageserver signals shutdown, current sync task gets finished and the loop exists.
+//!
+//! Currently there's no other way to download a remote relish if it was not downloaded after initial remote storage files check.
+//! This is a subject to change in the near future, but requires more changes to [`crate::tenant_mgr`] before it can happen.
+
+use std::{
+    cmp::Ordering,
+    collections::{hash_map, BinaryHeap, HashMap, HashSet},
+    path::{Path, PathBuf},
+    sync::{Arc, Mutex},
+    thread,
+    time::Duration,
+};
+
+use anyhow::{ensure, Context};
+use futures::stream::{FuturesUnordered, StreamExt};
+use lazy_static::lazy_static;
+use tokio::{sync::Semaphore, time::Instant};
+use tracing::*;
+
+use super::{RelishStorage, RemoteRelishInfo};
+use crate::{
+    layered_repository::{metadata_path, TimelineMetadata},
+    tenant_mgr::register_relish_download,
+    PageServerConf,
+};
+use zenith_metrics::{register_histogram_vec, register_int_gauge, HistogramVec, IntGauge};
+use zenith_utils::{
+    lsn::Lsn,
+    zid::{ZTenantId, ZTimelineId},
+};
+
+lazy_static! {
+    static ref REMAINING_SYNC_ITEMS: IntGauge = register_int_gauge!(
+        "pageserver_backup_remaining_sync_items",
+        "Number of storage sync items left in the queue"
+    )
+    .expect("failed to register pageserver backup remaining sync items int gauge");
+    static ref IMAGE_SYNC_TIME: HistogramVec = register_histogram_vec!(
+        "pageserver_backup_image_sync_time",
+        "Time took to synchronize (download or upload) a whole pageserver image. \
+        Grouped by `operation_kind` (upload|download) and `status` (success|failure)",
+        &["operation_kind", "status"],
+        vec![
+            0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1.0, 1.5, 2.0, 2.5, 3.0, 4.0, 5.0, 6.0, 7.0,
+            8.0, 9.0, 10.0, 12.5, 15.0, 17.5, 20.0
+        ]
+    )
+    .expect("failed to register pageserver image sync time histogram vec");
+}
+
+lazy_static! {
+    static ref SYNC_QUEUE: Mutex<BinaryHeap<SyncTask>> = Mutex::new(BinaryHeap::new());
```

Comment:
I'm thinking, why would I need `Mutex` here at all.
Would it be better to take an atomic `SegQueue` from https://github.com/crossbeam-rs/crossbeam/tree/master/crossbeam-queue and use that instead?

Reply:
I will note this scheme doesn't seem very fair, I think some tenant timeline pairs would always jump to the front. I think we'd want FIFO behavior. I'd consider reworking this (in a separate PR after merged) to use three separate queues+condvar or something, exact data structure details can be hashed out somewhere else.